enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,168 @@
+last-updated: 2023-02-15
+tracking-link:
+  - https://issues.redhat.com/browse/ETCD-391
+---
+
+# Periodic etcd Defragmentation in MicroShift
+
+## Summary
+
+This enhancement proposes adding a control loop to automatically and periodically run the etcd defragment commands.
+
+## Motivation
+
+Similar to a spinning-platter harddrive, etcd, through subsequent object creates and deletes, will have its internal database become fragmented. This causes etcd to use more space in memory and on disk than it actually has "in-use".
+Periodic defragmentation (along with compacation invoked by the API Server) is the way to ensure that etcd does not run out of space (hit the maximum-allowed database size of 8GB).
+
+Etcd running in the Openshift Container Platform (OCP) is automatically and periodically defragemented by the Cluster Etcd Operator (CEO). Because operators in Microshift are to be avoided, an alternative control loop for etcd defragmentation needs to be selected.
+
+### User Stories
+
+1. "As a Microshift Device Administrator, I want etcd to automatically defragment itself so that it doesn't run out of space, or use more space than it needs."
+
+### Goals
+
+- Etcd is automatically and periodically defragmented by a control loop that fits into the Microshift paradigm.
+
+### Non-Goals
+
+- Replicate other aspects/control loops of the OCP CEO.
+
+## Proposal
+
+All the following solutions will periodically calculate the "fragmented percentage" (db size on disk vs db in use). The frequency of this calculation and the threshold which the defragmentation occurs could be configurable (with defaults taken from OCP) by a Microshift Administrator so they can control when the defragmentation occurs.
+This configuration should also have the option to turn defragmentation completely off so that if an Administrator wants to opt-out of the potential for short disruptions, they have the ability to do so with the knowledge that etcd could run out of space during execution.
+For short-lived, light workload clusters, turning it off completely is likely not to cause problems.
+
+Additionally, it may be worth adding a pure periodic defragmentation so that an Administrator could force a defragmentation when load on the system is known to be low. 
+
+These two approaches could be combined so that the pure periodic defragmentation handles the normal day-to-day activity, while the threshold defragmentation is there if there are any sudden spikes in fragementation (caused by heavy usage).
+
+There is also the option of running defragmentation at Microshift startup (after etcd is alive and well). This would likely make the pure periodic redundant; however, we should still have the inflight threshold detection as that could help us avoid an out of space issue while the cluster is running.
+
+### Defragmentation from a goroutine in Microshift
+This solution would launch a goroutine, inside the Microshift binary, that would wait for the condition(s) to run defragmentation.
```

Comment:
would having this in `microshift-etcd` binary make sense?

Reply:
Could make sense to put it there, that way we group etcd-related tasks together. Especially because we have the server start notifier, we could put the go routine start right after that notification.

Reply:
I like the idea of it being self-managing.